object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -1,31 +1,176 @@
+    throw new Error("Do not instantiate");
   }
 
-  protected boolean evaluate(String expression) {
+  /** Construct a solver to use when comparing predicates. Each query should use a new solver. */
+  private static SolverContext setupSolver() {
+    Configuration config = Configuration.defaultConfiguration();
+    LogManager log = LogManager.createNullLogManager();
+    ShutdownNotifier notifier = ShutdownManager.create().getNotifier();
+    try {
+      return SolverContextFactory.createSolverContext(
+          config, log, notifier, SolverContextFactory.Solvers.SMTINTERPOL);
+    } catch (InvalidConfigurationException e) {
+      throw new BugInCF("Could not initialize solver.");
+    }
+  }
+
+  /**
+   * Return true iff the boolean formula "lhs &rarr; rhs" is valid, treating all variables as
+   * uninterpreted.
+   *
+   * @param lhs a Java boolean expression containing only variables, {@literal &&}, ||, !, and ()
+   * @param rhs a Java boolean expression containing only variables, {@literal &&}, ||, !, and ()
+   * @return true iff lhs implies rhs
+   * @throws UnsupportedOperationException if the input contains an unexpected kind of expression
+   */
+  public static boolean implies(final String lhs, final String rhs) {
+
+    // General approach: parse each formula into an AST, then convert each AST into a boolean
+    // formula that can be passed to a solver, and then construct a query to check if "not (lhs ->
+    // rhs)" is unsatisfiable. The result is the result of that query.
+
+    // set up solver
+    SolverContext context = setupSolver();
+
+    BooleanFormulaManager booleanFormulaManager =
+        context.getFormulaManager().getBooleanFormulaManager();
+
+    // parse the formulas into the solver's format
+    BooleanFormula lhsBool = formulaStringToBooleanFormula(lhs, booleanFormulaManager);
+    BooleanFormula rhsBool = formulaStringToBooleanFormula(rhs, booleanFormulaManager);
 
-    for (String cmMethod : cmMethods) {
-      expression = expression.replaceAll(cmMethod, "true");
+    BooleanFormula satQuery =
+        booleanFormulaManager.not(booleanFormulaManager.implication(lhsBool, rhsBool));
+    ProverEnvironment prover =
+        context.newProverEnvironment(SolverContext.ProverOptions.GENERATE_MODELS);
+
+    try {
+      prover.addConstraint(satQuery);
+    } catch (InterruptedException e) {
+      return false;
+    }
+
+    try {
+      return prover.isUnsat();
+    } catch (Exception e) {
+      return false;
     }
+  }
+
+  /** Converts a Java boolean expression into a boolean formula in the SMT solver's format. */
+  private static BooleanFormula formulaStringToBooleanFormula(
+      String formula, BooleanFormulaManager booleanFormulaManager) {
+
+    String classWithFormula = "class DUMMY { boolean DUMMY() { return " + formula + "; } }";
 
-    expression = expression.replaceAll("((?!true)[a-zA-Z0-9])+", "false");
+    CompilationUnit ast = parser.parse(classWithFormula).getResult().orElse(null);
```

Comment:
Rather than the next 10 lines, use `StaticJavaParser.parseExpression`.

Reply:
I had apparently overlooked that method. Thanks - this code is now much cleaner!